pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+    number of these :py:class:`Experiment` objects will be created and grouped into
+    an :py:class:`ExperimentSuite`.
+    """
+    in_operator: PauliTerm
+    out_operator: PauliTerm
+
+    def __str__(self):
+        return f'{self.in_operator}\u2192{self.out_operator}'
+
+    def __repr__(self):
+        return f'Experiment[{self}]'
+
+    def serializable(self):
+        return str(self)
+
+    @classmethod
+    def from_str(cls, s: str):
+        """The opposite of str(exp)"""
+        instr, outstr = s.split('\u2192')
+        return cls(in_operator=PauliTerm.from_str(instr),
+                   out_operator=PauliTerm.from_str(outstr))
+
+
+def _abbrev_program(program: Program, abbrev_after=10):
+    program_lines = program.out().splitlines()
+    if abbrev_after is not None and len(program_lines) > abbrev_after:
+        first_n = abbrev_after // 2
+        last_n = abbrev_after - first_n
+        excluded = len(program_lines) - abbrev_after
+        program_lines = (program_lines[:first_n] + [f'... {excluded} hidden ...']
+                         + program_lines[-last_n:])
+
+    return '; '.join(program_lines)
+
+
+class ExperimentSuite:
+    """
+    A whole set of tomography-like experiments
+
+    Many near-term quantum algorithms involve measuring an ansatz state with respect to
+    many different observables. This class contains all (in_pauli, out_pauli) pairs for
+    a given ansatz preparation program.
+
+    Experiments belonging to a shared tensor product basis (TPB) can (optionally) be estimated
+    simultaneously. Therefore, this class is backed by a list of list of Experiments.
+    Experiments sharing an inner list will be estimated simultaneously. If you don't want this,
+    provide a list of length-1-lists. As a convenience, if you pass a 1D list to the constructor
+    it will expand it to a list of length-1-lists.
+
+    Please see :py:func:`group_experiments` for a function that will automatically process an
+    ExperimentSuite to group Experiments sharing a TPB.
+    """
+
+    def __init__(self,
+                 experiments: Union[List[Experiment], List[List[Experiment]]],
+                 program: Program,
+                 qubits: List[int]):
+        if len(experiments) == 0:
+            experiments = []
+        else:
+            if isinstance(experiments[0], Experiment):
+                # convenience wrapping in lists of length 1
+                experiments = [[exp] for exp in experiments]
+
+        self._experiments = experiments  # type: List[List[Experiment]]
+        self.program = program
+        self.qubits = qubits
+
+    def __len__(self):
+        return len(self._experiments)
+
+    def __getitem__(self, item):
+        return self._experiments[item]
+
+    def __setitem__(self, key, value):
+        self._experiments[key] = value
+
+    def __delitem__(self, key):
+        self._experiments.__delitem__(key)
+
+    def __iter__(self):
```

Comment:
Just noting that, using this definition of the `__iter__` function (and similarly, the `__reversed__` function), if we loop over an ExperimentSuite instance, we have lists, not experiments. Is this the desired functionality?

Reply:
for better or worse, yes

Reply:
I think this is fine.